### DISCUSSION TITLE
Knex.js and Postgres with Next.js API Routes

### DISCUSSION THREAD

**Discussion by barelyhuman** [vercel/next.js#13556] [Help] [answered]
Knex keeps creating a lot of connections (because every request creates a new connection) and idles out the query even when the request has been full filled, destroying the connection is one solution but I'd like to know what other people are using.

**Comment by apnerve** [comment]
Does this happen during local development or in production? I face the same issue with prisma during development but in production, it is doesn't

**Reply by barelyhuman** [reply]
It happens both in prod and local since the connections are stuck on idle, as in the question above, I do have it managed by destroying the connection after completing the request but it feels hacky and would prefer a better solution, so wanted to know if any other next.js users had figured out a better way

**Reply by juanbzz** [reply]
> I do have it managed by destroying the connection after completing the request but it feels hacky and would prefer a better solution

Some ORMs handle the issue by doing what you are doing, closing the connection after the request. See https://github.com/sequelize/sequelize/issues/8468

**Reply by barelyhuman** [reply]
Guess I’m gonna stay with my solution then.

**Comment by barelyhuman** [comment, marked as answer]
The solution remains to be to destroy the knex connection after the request has been fulfilled.

**Reply by kennethacohen** [reply]
@barelyhuman any chance you can share an example here? only fix I have been able to get is to store a global connection, which felt wrong...would appreciate any help 😄

**Reply by barelyhuman** [reply]
@kennethacohen 
https://reaper.im/blog/nextjs-pg-knexjs


I guess writing about it wasn’t a bad idea 😂

**Reply by rhazegh** [reply]
@barelyhuman Thank you! This solution works perfectly in `pages/api/my-end-point.js` 🎉

But when I do `import connectionHandler from 'connection-handler';` inside `pages/my-page.js` I get a compile error:
```
error - ./node_modules/knex/lib/migrations/util/fs.js:1:0
Module not found: Can't resolve 'fs'
```

Is there a way to use the `connection-handler` inside a regular page (not an API route)?

**Reply by barelyhuman** [reply]
@rhazegh knex requires it to be a node env and not a browser env , you can use `getServerSideProps` to use knex on a regular page if the use case involves rendering the page with data from the server side itself.

**Reply by rhazegh** [reply]
@barelyhuman Thank you. Yeah, that's what I ended up doing. I would like to check for user authorization and redirect to another page if they are not authorized. I understand the other option is doing the authorization check on the client side but that has it's own problems.
